crates/types/src/config/worker.rs

```diff
@@ -124,10 +137,86 @@ impl Default for WorkerOptions {
             max_command_batch_size: NonZeroUsize::new(32).expect("Non zero number"),
             snapshots: SnapshotsOptions::default(),
             trim_delay_interval: None,
+            durability_mode: None,
+            experimental_partition_driven_log_trimming: false,
         }
     }
 }
 
+#[serde_as]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, derive_more::Display, Serialize, Deserialize)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "schemars", schemars(rename = "DurabilityMode"))]
+#[serde(rename_all = "kebab-case")]
+pub enum DurabilityMode {
+    /// This disables durability tracking and trimming completely.
+    ///
+    /// Trims and snapshots are still possible if performed manually or by an external
+    /// component.
+    None,
+
+    /// In this mode, a partition is considered durable when its state can be restored from
+    /// any of members of the replica-set as well as the latest snapshot.
+    ///
+    /// In other words, do not trim unless **all** replicas cover this Lsn, **and** the snapshot.
+    ///
+    /// [requires snapshot repository]
+    /// DurabilityPoint = Min(Min(ReplicaSetDurablePoints), SnapshotDurablePoint)
+    SnapshotAndReplicaSet,
+
+    // A partition is fully durable if the _entire_ replica-set is durable at this Lsn
+    // **or** if a snapshot is available.
+    //
+    // Do not trim unless the Lsn is covered by a snapshot, **or** by _all_ the replicas
+    // in the replica-set.
+    //
+    // DurabilityPoint = Max(Min(ReplicaSetDurablePoints), SnapshotDurablePoint)
+    //
+    // [Requires node-to-on-node sharing of ad-hoc snapshots] so it's commented until support is
+    // added
+    // SnapshotOrReplicaSet,
+    // -----
+    /// In this mode, a partition is considered durable when its state can be restored from
+    /// the snapshot and at least a single replica.
+    ///
+    /// Do not trim unless the Lsn is covered (durably) by _any_ of the replicas **and** by
+    /// the snapshot. Gives weight to snapshots over the durability of the replica-set but
+    /// without ignoring the replica-set completely.
+    ///
+    /// In practice, this means that after a snapshot has been created on the leader, the
+    /// system will wait for the nearest memtable flush that cover this Lsn before considering
+    /// this Lsn for trimming. If the leader crashes before the memtable flush, we are confident
```

Comment:
Wouldn't creating a snapshot trigger a flush operation on the leader (at least right now `Checkpoint::export_column_family`) and thereby automatically fulfill the requirement? Or would the leader be excluded from the `ReplicaSetDurablePoints`?

Reply:
>Wouldn't creating a snapshot trigger a flush operation on the leader (at least right now Checkpoint::export_column_family) and thereby automatically fulfill the requirement?

Yes, that's correct, but that won't hold true if if we added support for exporting without flush. My hope is that we'd be able to do that soon to avoid the extra flush unless the snapshot min_applied_lsn > durable_lsn.

Even now, our view of the view in ReplicaSetDurablePoints might be a little behind what happened during the export  since this is updated by PP's `status_update_timer`.